src/redux/reducers/genes/genesExpressionLoaded.js

```diff
@@ -1,34 +1,26 @@
-import _ from 'lodash';
-import { initialViewState } from './initialState';
-
-const genesExpressionLoaded = (state, action) => {
-  const upperCaseArray = (array) => (array?.map((element) => element.toUpperCase()));
-
-  const {
-    data, componentUuid, genes,
-    loadingStatus = _.difference(upperCaseArray(state.expression.loading), upperCaseArray(genes)),
-  } = action.payload;
-  return {
-    ...state,
-    expression: {
-      ...state.expression,
-      views: {
-        ...state.expression.views,
-        [componentUuid]: {
-          ...initialViewState,
-          ...state.expression.views[componentUuid],
-          fetching: false,
-          error: false,
-          data: genes,
-        },
-      },
-      data: {
-        ...state.expression.data,
-        ...data,
-      },
-      loading: loadingStatus,
-    },
-  };
-};
+/* eslint-disable no-param-reassign */
+import produce from 'immer';
+
+import { calculateZScore } from 'utils/postRequestProcessing';
+import initialState from 'redux/reducers/genes/initialState';
+import { difference as _difference } from 'lodash';
+
+const upperCaseArray = (array) => (array?.map((element) => element.toUpperCase()));
+
+const genesExpressionLoaded = produce((draft, action) => {
+  const { data, componentUuid, genes } = action.payload;
+
+  let { loadingStatus } = action.payload;
+
+  if (loadingStatus === undefined) {
+    loadingStatus = _difference(upperCaseArray(draft.expression.loading), upperCaseArray(genes));
+  }
+
+  const dataWithZScore = calculateZScore(data);
```

Comment:
why isn't `calculateZScore` function get called from within `fetchGeneExpressionWork` when the gene expression is requested?

Reply:
Having to remember to call `calculateZScore` as well as any other post-request processing tasks in individual actions is prone to errors and forgetting. It will be ideal to call them from one place for all gene expression related stuff and not have to worry about calling it in individual actions.

Reply:
`fetchGeneExpressionWork` currently returns 2 work types: load gene expression and load marker genes. There is a difference between the data that's returned from these 2 work requests. Unless we want to handle differences in the strutcure of the returned body, I think it's best to handle this just before storing the data in redux.